lib/attack.go

```diff
@@ -223,7 +223,61 @@ func Client(c *http.Client) func(*Attacker) {
 	return func(a *Attacker) { a.client = *c }
 }
 
-// A Rate of hits during an Attack.
+// A Rater defines the rate of hits during an Attack by
+// returning the duration an Attacker should wait until hiting
+// the next Target. When a negative value time is returned,
+// an Attacker will stop.
+type Rater interface {
+	Wait(began, now time.Time) time.Duration
+}
+
+// A RaterFunc is a function adapter type that implements the
+// Rater interface.
+type RaterFunc func(began, now time.Time) time.Duration
+
+// Wait implements the Rater interface.
+func (f RaterFunc) Wait(began, now time.Time) time.Duration {
+	return f(began, now)
+}
+
+// A FixedRater is a Rater that produces hits at a fixed rate
+// for a given duration. It is not safe for concurrent use.
+type FixedRater struct {
+	rate  Rate
+	hits  uint64
+	count uint64
+}
+
+// NewFixedRater returns a new FixedRater with the given Rate and duration.
+// When duration is zero, the rater will never stop.
+func NewFixedRater(rate Rate, du time.Duration) *FixedRater {
+	r := FixedRater{rate: rate}
+	if du != 0 && !rate.IsZero() {
```

Comment:
If `rate.IsZero` returns true, this should probably return an error and a nil `*FixedRater`. Otherwise the implementation of `Wait` as it is now could panic with integer-divide-by-zero if `rate.Freq` is zero.

Reply:
This also point out a dent in tests coverage ;)

Reply:
Good catch. What do you think of using a zero valued Rate as meaning no pacing at all? Just go as fast as you can.